examples/function/asyncio-get/executorpool.py

```diff
@@ -0,0 +1,132 @@
+
+
+class ExecutorPool:
+    """
+       A class used to create an Executor Pool using the "concurrent.futures" library
+       Currently ThreadPool and ProcessPool are supported
+
+       ...
+
+       Attributes
+       ----------
+       executor : ThreadPoolExecutor or ProcessPoolExecutor
+           a formatted string to print out what the animal says
+       loop : asyncio event loop
+           the event loop on which the tasks should be submitted
+
+       Methods
+       -------
+       async submit(func, *args)
+           Submit the given function to the executor, returns a futures object
+
+       async gather(tasks, return_exceptions, return_when)
+           Wait for the coroutines to execute and gather the results
+
+       set_loop(event_loop)
+           Update the event loop associated with the executor
+
+       close(wait)
+            Shutdown the executor
+       """
+
+    def __init__(self, exec_type="thread", max_workers=None, loop=None):
+        self.loop = loop or asyncio.get_event_loop()
+        if exec_type == "thread":
+            self.executor = concurrent.futures.ThreadPoolExecutor(max_workers=max_workers)
+        elif exec_type == "process":
+            self.executor = concurrent.futures.ProcessPoolExecutor(max_workers=max_workers)
+
+    async def submit(self, func, *args):
+        return self.loop.run_in_executor(self.executor, func, args)
+
+    async def gather(self, tasks, return_exceptions=False, return_when="ALL_COMPLETED"):
+        completed, pending = await asyncio.wait(tasks, return_when=return_when)
+        results = [t.result() for t in completed]
+        if return_exceptions:
+            results += [t.exception() for t in completed if t.exception()]
+        # self.tasks = []
+        return results
+
+    def set_loop(self, event_loop):
+        self.loop = event_loop
+
+    def close(self, wait=True):
+        self.executor.shutdown(wait)
+
+
+"""
+Invoke a I/O blocking call using http requests
+"""
+
+
+def blocking_call(url):
+    url = url[0]
+    res = requests.get(url)
+    try:
+        res = res.json()
+        return res["message"]
+    except Exception as e:
+        _LOGGER.error("HTTP request error: %s", e)
+        return "Error"
+
+
+"""
+Invoke a CPU blocking call by computing factorial of a given number
+"""
+
+
+def blocks(n):
+    # log = logging.getLogger('blocks({})'.format(n))
+    # log.info('running')
+    n = n[0]
+    fact = 1
+    for i in range(1, n + 1):
+        fact = fact * i
+    return fact
+
+
+"""
+handler function for executing operations on the messages received by the reduce vertex
+Here the function is used to execute certain blocking operations to demonstrate the use of 
+asyncio with ThreadPool and ProcessPool Executor
+"""
+
+
+async def reduce_handler(key: str, datums: AsyncIterable[Datum], md: Metadata) -> Messages:
+    interval_window = md.interval_window
+    tasks = []
+    threadPool.set_loop(asyncio.get_event_loop())
```

Comment:
since this function requires threadPool, its a good idea to wrap these functions inside of a class and pass in the threadpool reference explicitly.